### ISSUE TITLE
Alternatively, enable callable class constructors

### ISSUE THREAD

**Issue by allenwb** [jasnell/proposal-construct#2]
Won't you be able to deal with old style subclass "super" constructor calls if ES class constructors were permitted to be [[Call]] invoked rather than throwing when called?

Callable class constructors were part of the original ES class design and have only been deferred because of ergonomic concerns.  Perhaps this use case is what is needed to push them over the hump.

**Comment by jasnell** [comment]
Yes, that would do nicely. There would obviously need to be some tweaking done with `super()` but allowing class constructors to be called would solve it nicely.

**Comment by davidmarkclements** [comment]
I agree - instead of adding more syntax, is it possible to just tweak the behaviour - in this case making classes callable won't cause any breakage so it's a minor adjustment without additional language surface area

**Comment by loganfsmyth** [comment]
My biggest concern here is that there isn't a clear way to generalize this behavior across all constructors. It seems workable for the specific case you're looking to resolve, but for the general case there are tons of unanswered questions.

Making constructors callable would only work with constructors that don't need to be responsible for constructing their own objects. Any constructor that extends a type that relies on creating an exotic object like `Array` or `Map` or `Error` or many others, would either have to fail, or an entirely new process would need to be created so that the proper exotic object could be initialized by the caller before calling the constructor.

Alternatively if the goal was to make them callable _without_ passing in an initial value for `this` so the constructor could build them, you run into issues of the constructor not having a `new.target` value to tell it what prototype to use when constructing the object.

**Comment by ljharb** [comment]
It would have to be opt-in. It's important that existing classes continue to throw instead of being callable.

Classes that do opt in could, perhaps, do so knowing that `new.target` might be `undefined`.

**Comment by lifaon74** [comment]
Well, personally I really feel a need to have access to the constructor of a class as a pure function (just like ES5 classes) => calling the constructor with a different **this** and a list of dynamic arguments (thanks Function.apply). Allow a precise `toString()` to get the content, etc...

Currently : 
```js
class A {
  constructor() {
    console.log('init');
  }
}

function B() {
  console.log('init');
}

console.log(A.prototype.constructor); // print the whole class
console.log(B.prototype.constructor) // print the function
// So A.call is not valid but B.call is ! And I feel that this proposal tries to fix this difference.
```
When creating a es6 class there should be a way to access to the constructor as a function (to allow apply, call, etc...). The current proposal only propose a `construct` but for me it only cover a portion of the problem, not the whole.
Maybe we could have something like  : `A.prototype[Symbol.constructor]` which returns the constructor as a function, not the class itself, allows to edit the constructor and to call it directly without new.

**Comment by ljharb** [comment]
`Reflect.construct` might be helpful.

**Comment by lifaon74** [comment]
I know about `Reflect.construct` but the thing is that it only **construct**. You have no access to the constructor function (name, arguments length, toString, call, apply, etc...). This is what the author of this proposal and me are pointing.
